### PR TITLE
Make addresses append only

### PR DESCRIPTION
**Pull request by zancas**

### CODE REVIEW

**Review by fluidvanadium — DISMISSED** [review]

File: zingolib/src/wallet/keys/unified.rs [hunk 1]
```diff
@@ -167,52 +183,78 @@ impl WalletCapability {
...
+                    .find_address(sapling_diversifier_index)
+                    .expect("Diversifier index overflow");
+                sapling_diversifier_index
+                    .increment()
+                    .expect("diversifier index overflow");
+                if count == self.addresses.len() {
+                    break;
+                }
+                count += 1;
+            }
```
Comment (AloeareV) [inline zingolib/src/wallet/keys/unified.rs:225]: I think a for loop would be cleaner here..something like this?
```suggestion
            for i in 0..self.addresses.len() {
                sapling_diversifier_index
                    .increment()
                    .expect("diversifier index overflow");
            }
            Some(fvk
                .find_address(sapling_diversifier_index)
                .expect("Diversifier index overflow"));
```

File: zingolib/src/wallet/keys/unified.rs [hunk 2]
```diff
@@ -230,11 +272,22 @@ impl WalletCapability {
...
+                return Err(
+                    "Invalid receivers requested! At least one of sapling or orchard required"
+                        .to_string(),
+                );
+            }
+        };
         self.addresses.push(ua.clone());
+        assert_eq!(self.addresses.len(), previous_num_addresses + 1);
+        self.addresses_write_lock
+            .swap(false, atomic::Ordering::Release);
```
Comment (AloeareV) [inline zingolib/src/wallet/keys/unified.rs:290]: Badass!!

**Review by AloeareV — CHANGES_REQUESTED** [review]
Awesome! Just a suggested loop tidying

File: zingolib/src/wallet/keys/unified.rs (hunk 1 above)
Comment (zancas) [reply zingolib/src/wallet/keys/unified.rs:225]: Did you test this suggestion against the `determinism` test?

**Comment by zancas** [comment]
I've converted this to draft on @fluidvanadium 's recommendation that we benchmark before merge.

File: zingolib/src/wallet/keys/unified.rs (hunk 1 above)
Comment (AloeareV) [reply zingolib/src/wallet/keys/unified.rs:225]: Talked through this code line-by-line with @fluidvanadium. I revert my suggestion, your fancy loop is nessecary.

**Comment by AloeareV** [comment]
> vzing-0 3 0-462-g10a6733b_sync_duration_annotation
> 
> I think this data, makes this PR no-longer-draft.

Is there an easy way to compare this to dev benchmarks ran under the same circumstances?

**Review by AloeareV — APPROVED** [review]

**Comment by zancas** [comment]
mob-release-1 2 5_126_sync_duration_annotation

This is the same benchmarks, run against the version of zingolib that's compiled into our last release.

**Review by fluidvanadium — APPROVED** [review]

**Comment by zancas** [comment]
mob-release-1 2 5_126-46-g65067dad_sync_duration_annotation

@AloeareV 

Here's the benchmark for the commit that was dev before this merged.

**Pull request merged**